File: 3k/1s/cpp/conversation/Person.cpp
#include <iostream>
#include <string>
#include <windows.h>

using namespace std;

class Person {
private:
    /** Имя **/
    char *name;

    /** Возраст **/
    int age;

protected:
    void init(char* newName, int newAge) {
        name = new char[strlen(newName) + 1];
        strcpy(name, newName);

        age = newAge;
    }

public:

    const void fullHello(Person* &personTwo)
    {
        cout << name << ": ";
        Hello(personTwo);
    }

    virtual const void Hello(Person* &personTwo) = 0;

    virtual const char* getTypeText() {
        return "Человек";
    }

    char* getName()
    {
        return name;
    }

    void about()
    {
        cout << "Меня зовут " << name << ", мой возраст " << age << " " << viewAge() << ", я " << getTypeText();
    }

    int getAge()
    {
        return age;
    }

    /**
     * Выводим: год/лет/года
     */
    char* viewAge()
    {
        switch(age % 10)
        {
            case 1:
                return "год";
                break;
            case 2: case 3: case 4:
                return "года";
                break;
            default:
                return "лет";
                break;
        }
    }
};

File: 3k/2s/java/5. Conversation/src/com/company/Person.java
package com.company;

public abstract class Person implements PersonInterface
{
    // Мужской пол
    public static final int SEX_MALE = 1;

    // Женский пол
    public static final int SEX_FEMALE = 2;

    /** Имя **/
    private String name;

    /** Возраст **/
    private int age;

    /** Пол **/
    private int sex;

    public Person(String newName, int newAge, int newSex) {
        name = newName;
        age = newAge;
        sex = newSex;
    }

    abstract protected void hello(Person personTwo);

    abstract public String getTypeText();

    /**
     * Получить пол
     */
    public int getSex()
    {
        return sex;
    }

    /**
     * Получить пол текстом
     */
    public String getSexText()
    {
        return (sex == 1) ? "мужской" : "женский";
    }

    public void fullHello(Person personTwo)
    {
        System.out.print(name + ": ");

        hello(personTwo);
    }

    /**
     * Получить имя персоны
     */
    public String getName()
    {
        return name;
    }

    /**
     * Вывод информации о персоне
     */
    public void about()
    {
        System.out.println("Меня зовут: " + name);
        System.out.println("Мой возраст: " + getAgeText());
        System.out.println("Мой пол: " + getSexText());
        System.out.println("Я - " + getTypeText());
    }

    /**
     * Получить возраст
     */
    public int getAge()
    {
        return age;
    }

    /**
     * Получить возраст с корректным склонением
     */
    public String getAgeText()
    {
        switch(age % 10)
        {
            case 1:
                return age + " год";
            case 2: case 3: case 4:
                return age + " года";
            default:
                return age + " лет";
        }
    }

    protected String getTypeTextSuffix()
    {
        return sex == SEX_FEMALE ? "ка" : "";
    }
}
